File: C++/source/GameBoard.cc
//
// Created by liamg on 24/10/2025.
//

#include "../header/GameBoard.h"

#include <algorithm>

using namespace std;

struct BoardFeature {
  int start;
  int end;
};

constexpr BoardFeature LADDERS[] = {
  {.start = 2, .end = 9},
  {.start = 11, .end = 28},
  {.start = 19, .end = 31}
};

constexpr BoardFeature SNAKES[] = {
  {.start = 12, .end = 4},
  {.start = 27, .end = 12},
  {.start = 30, .end = 20}
};

bool GameBoard::isInSnakes(const int position) {
  return ranges::any_of(SNAKES, [&](const BoardFeature& feature) {
    return feature.start == position;
  });
}

int GameBoard::getSnakeEnd(const int startPos) {
  for (auto [start, end] : SNAKES) if (start == startPos) return end;

  return startPos;
}

bool GameBoard::isInLadders(const int position) {
  return ranges::any_of(LADDERS, [&](const BoardFeature& feature) {
    return feature.start == position;
  });
}

int GameBoard::getLadderEnd(const int startPos) {
  for (auto [start, end] : LADDERS) {
    if (start == startPos) return end;
  }

  return startPos;
}

File: scala/src/main/scala/GameBoard.scala
object GameBoard {
  
  val BOARD_SIZE = 35

  private val LADDERS = Array(
    new BoardFeature(2, 9),
    new BoardFeature(11, 28),
    new BoardFeature(19, 31)
  )

  private val SNAKES = Array(
    new BoardFeature(12, 4),
    new BoardFeature(27, 12),
    new BoardFeature(30, 20)
  )

  def isInSnakes(position : Int) : Boolean = {
    SNAKES.exists(s => s.getStart == position)
  }

  def getSnakeEnd(startPos : Int) : Int = {
    SNAKES.find(s => s.getStart == startPos).get.getEnd
  }

  def isInLadders(position: Int): Boolean = {
    LADDERS.exists(l => l.getStart == position)
  }

  def getLadderEnd(startPos: Int): Int = {
    LADDERS.find(l => l.getStart == startPos).get.getEnd
  }
}
